src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractTranslator.java

```diff
@@ -331,6 +333,59 @@ public String translate(Update u) {
         return join(temp, " ");
     }
 
+    /**
+     * Translates When.
+     *
+     * @param when a when.
+     * @return when translation.
+     */
+    public String translate(final When when) {
+        inject(when.condition);
+        inject(when.action);
+
+        return String.format("WHEN %s THEN %s",
+                             when.condition.translate(),
+                             when.action.translate());
+    }
+
+    /**
+     * Translates Case.
+     *
+     * @param aCase a case.
+     * @return case translation.
+     */
+    public String translate(final Case aCase) {
+        String elseString = "";
+        if (aCase.getFalseAction() != null) {
+            inject(aCase.getFalseAction());
+            elseString = String.format("ELSE %s", aCase.getFalseAction().translate());
+        }
+        return String.format("CASE %s %s END",
+                             caseBuilder(aCase).toString(),
+                             elseString);
+    }
+
+    /**
+     * Translates a when inside a case expression.
+     *
+     * @param aCase a case expression.
+     * @return a translated when expression.
+     */
+    private StringBuilder caseBuilder(final Case aCase) {
+        List<When> whens = aCase.whens;
+
+        inject(whens);
+        StringBuilder builder = new StringBuilder();
+
+        for (int i = 0; i < whens.size(); i++) {
```

Comment:
I believe this would be cleaner if using a stream, something like:

```
whens.stream()
  .peek(this::inject)
  .map(When::translate)
  .collect(Collectors.joining(" "));
```
I did not test it, but it makes conceptually sense.
Maybe after this you don't event need this method and move this to the one above.

That said, functionally it is the same, so it is your call how you design the code.